lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp

```diff
@@ -364,6 +338,130 @@ void FirRegLower::initialise(OpBuilder &regBuilder, OpBuilder &initBuilder,
   recurse(type, "");
 }
 
+bool FirRegLower::isInitialisePreset(sv::RegOp reg) {
+  return module->hasAttr("firrtl.random_init_width") &&
+         reg->hasAttr("firrtl.random_init_start") &&
+         reg->hasAttr("firrtl.random_init_end");
+}
+
+void FirRegLower::initialisePreset(OpBuilder &regBuilder,
+                                   OpBuilder &initBuilder, sv::RegOp reg) {
+  // Extract the required random initialisation width and compute the number of
+  // `RANDOM calls required to fill it up, as well as the final register width.
+  auto randomInitWidthAttr =
+      module->getAttrOfType<IntegerAttr>("firrtl.random_init_width");
+  assert(randomInitWidthAttr &&
+         "firrtl.random_init_width required for preset initialisation");
+
+  uint64_t randomInitWidth = randomInitWidthAttr.getUInt();
+  assert(randomInitWidth > 0 &&
+         "random initialization width should be non-zero");
+  uint64_t numRandomSources = llvm::divideCeil(randomInitWidth, randomWidth);
+  uint64_t randomRegWidth = randomWidth * numRandomSources;
+
+  // Only create the random register once.
+  if (presetRandomValue == nullptr) {
+    // Declare the random register.
+    auto randReg = regBuilder.create<sv::RegOp>(
+        reg.getLoc(), regBuilder.getIntegerType(randomRegWidth),
+        /*name=*/regBuilder.getStringAttr("_RANDOM"),
+        /*inner_sym=*/
+        regBuilder.getStringAttr(ns.newName(Twine("_RANDOM"))));
+
+    presetRandomValue = randReg;
+
+    SmallString<32> randomRegAssign("{{0}} = {");
+
+    // Fill the random register by concatenating calls to `RANDOM in a verbatim
+    // string.
+    for (uint64_t i = 0; i < numRandomSources; ++i) {
+      randomRegAssign.append("`RANDOM");
+      if (i < numRandomSources - 1)
+        randomRegAssign.append(",");
+      // Add a line break when line length gets close to 1000 characters.
+      if (i > 0 && i % 125 == 0)
+        randomRegAssign.append("\n");
+    }
+
+    randomRegAssign.append("};");
```

Comment:
Probably we can use `for` macro to initialize the random register in a similar way to HWMemSimImp(https://github.com/llvm/circt/blob/main/lib/Dialect/SV/Transforms/HWMemSimImpl.cpp#L418-L428)  in the future.

Reply:
Yeah I forgot about that one. I think it makes sense, but I also think I'll merge this as-is since it's been tested.

Reply:
Sounds good!